pkg/logs/decoder/decoder.go

```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
 			maxj = i + d.contentLenLimit
 		} else if inBuf[j] == '\n' {
 			d.lineBuffer.Write(inBuf[i:j])
-			d.sendLine()
-			i = j + 1 // +1 as we skip the `\n`
+			if !matchDockerHeader(d.lineBuffer, d.useDockerParser) {
+				d.sendLine()
+			} else {
+				d.lineBuffer.WriteByte(inBuf[j])
+
+			}
+			i = j + 1
 			maxj = i + d.contentLenLimit
 		}
 	}
 	d.lineBuffer.Write(inBuf[i:j])
 }
 
+var dockerHeaderLen = 8
+var dockerHeaderPrefixLen = 4
+var dockerHeaderPrefix1 = []byte{1, 0, 0, 0}
+var dockerHeaderPrefix2 = []byte{2, 0, 0, 0}
+
+// When a newline (in byte is 10) is matching, an additional check need to
+// be done to make sure this is not part of docker header.
+// [1|2 0 0 0 size1 size2 size3 size4], where size1 size2 size3 size4 can be
+// 10 in byte.
+// case [1|2 0 0 0 10 size2 size3 size4]
+// case [1|2 0 0 0 sizeNot10 10 size3 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 10 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 sizeNot10 10]
+// TODO: Refactor decoder to make this check specific to docker source
+func matchDockerHeader(buffer *bytes.Buffer, useDockerParser bool) bool {
+	if !useDockerParser {
+		return false
+	}
+	l := buffer.Len()
+	// len must be less than header length
```

Comment:
when you add a comment, you should explain why we do things, not what we do. What we do could be explain in the function description.